Update scala-cli.md (#2)
romanowski/improvement-proposals: content/scala-cli.md

## Summary

We propose to replace current scripts that are installed as "Scala": `scala`, `scalac` and `scaladoc` with Scala CLI - a batteries included tool to interact with Scala. Scala CLI brings all the features that the commands above provide and expand them with incremental compilation, dependency management, packaging and much more. 

## Motivation

The current default Scala scripts: `scala`, `scalac`, `scaladoc` are quite limited. Beside that, the `scala` command that starts the Scala REPL is hardly used by non-power users. 

> The current scripts are lacking basics features such as support for resolving dependencies, incremental compilation or support for outputs other than JVM. This forces any user that wants to do anything more then just basic things to learn and use SBT, Mill or an other build tool and that adds to the complexity of learning Scala. 

Reviewer: ```suggestion
The current scripts are lacking basic features such as support for resolving dependencies, incremental compilation or support for outputs other than JVM. This forces any user that wants to do anything more than just basic things to learn and use SBT, Mill or an other build tool and that adds to the complexity of learning Scala. 
```

> We observe that current state of tooling is Scala is limiting the creativity, with quite high cost to create e.g. an application or script with some dependencies that targets Node.js. Many Scala developers are not choosing Scala for their personal projects, scripts, or small applications and we believe that the complexity of setting up a build tool is one of the reasons. 

Reviewer: ```suggestion
We observe that the current state of tooling in Scala is limiting creativity, with quite a high cost to create e.g. an application or a script with some dependencies that target Node.js. Many Scala developers are not choosing Scala for their personal projects, scripts, or small applications and we believe that the complexity of setting up a build tool is one of the reasons. 
```

## Proposed solution

We propose to gradually replace the current `scala`, `scalac` and `scaladoc` commands by single `scala` command that under the hood will be `scala-cli`. We could also add wrapper scripts for `scalac` and `scaladoc` that will mimic the functionality that will use `scala-cli` under the hood. 

The complete set of `scala-cli` features can be found in [its documentation](https://scala-cli.virtuslab.org/docs/overview).

Scala CLI brings many features like testing, packaging, exporting to sbt / Mill or upcoming support for publishing micro-libraries. Initially, we may want to limit the set of features available in the `scala` command by default. Scala CLI is a relatively new project and we should battle-proof some of its features before we commit to support them as part of the offical `scala` command. 

> Scala CLI offers a [multiple native ways to be installed](https://scala-cli.virtuslab.org/install#advanced-installation) so most users should find a suitable method. We would like these packages to become the default `scala` packages in most repositories, often replacing existing `scala` packages.

Reviewer: ```suggestion
Scala CLI offers [multiple native ways to be installed](https://scala-cli.virtuslab.org/install#advanced-installation) so most users should find a suitable method. We would like these packages to become the default `scala` package in most repositories, often replacing existing `scala` packages.
```

### High-level overview

> Let us show a few examples where adopting Scala CLI as `scala` command would be a significant improvement ofer current scripts. For this, we have assumed a minial set of features. Each additionnal  Scala CLI feature included such as `package` would add more and more use cases.

Reviewer: ```suggestion
Let us show a few examples where adopting Scala CLI as `scala` command would be a significant improvement ofer current scripts. For this, we have assumed a minial set of features. Each additionnal  Scala CLI feature included, such as `package`, would add more and more use cases.
```

**Using REPL with a 3rd-party dependency**

Currently, to start a Scala REPL with a dependency on the class path, users need to resolve this dependency with all its transitive dependencies (coursier can help here) and pass those to the `scala` command using the `--cp` option. Alternatively, one can create an sbt project including a single dependency and use the `sbt console` task. Ammonite gives a better experience with its magic imports. 

With Scala CLI, starting a REPL with a given dependency is as simple as running:

```scala-cli repl --dep com.lihaoyi::os-lib:0.7.8```

> Compared to Ammonite, default Scala REPLs provided by Scala 2 and 3 - that Scala CLI uses by default - are somewhat limited. However Scala CLI also offers to start Ammonite instead of the default Scala REPL, by passing `--ammonite` (or `--amm`) option to `scala-cli repl`.

Reviewer: ```suggestion
Compared to Ammonite, default Scala REPLs provided by Scala 2 and 3 - that Scala CLI uses by default - are somewhat limited. However, Scala CLI also offers to start Ammonite instead of the default Scala REPL, by passing `--ammonite` (or `--amm`) option to `scala-cli repl`.
```

Additionally, `scala-cli repl` can also put code from given files / directories / snippets on the class path by just providing their locations as arguments. Running ```scala-cli repl foo.scala baz``` will compile code from `foo.scala` and the `baz` directory, and put their classes on the REPL class path (including their dependencies, scalac options etc. defined within those files).

> Compilation (and running scaladoc as well) benefit in the similar way from the ability manage dependencies.

Reviewer: ```suggestion
Compilation (and running scaladoc as well) benefit in the similar way from the ability to manage dependencies.
```

Author: Should even been "in *a* similar way"

> Compilation (and running scaladoc as well) benefit in the similar way from the ability to manage dependencies.

Author: ```suggestion
Compilation (and running scaladoc as well) benefit in a similar way from the ability to manage dependencies.
```

def foo = println("<here comes the buggy warning with Scala Native and os-lib>")
```

The snippet above when run with Scala CLI without any configuration provided will use Scala Native, resolve and include `os-lib` and provide `-Xfatal-warnings` to the compiler. Even things such as the runtime JVM version can be configured with using directives.

Moreover, Scala CLI provides the ability to run GitHub gists (including multi-file ones), and more.

** Watch mode **

> When working on a piece of code, it is often useful to have it compiled/run everytime the file is changed, and build tools offer a watch mode for that. This is how most people are using watch mode through a build tool. Scala CLI offers watch more for most commands (by using `--watch` / `-w` flags).

Reviewer: ```suggestion
When working on a piece of code, it is often useful to have it compiled/run everytime the file is changed, and build tools offer a watch mode for that. This is how most people are using watch mode through a build tool. Scala CLI offers a watch mode for most of its commands (by using `--watch` / `-w` flags).
```

### Specification

 In order to be able to expand the functionality of Scala CLI and yet use the same core to power the `scala` command, we propose to include both `scala` and `scala-cli` commands in the installation package. Scala CLI already has a feature to limit accessible sub-commands based the binary name (all sub-commands in `scala-cli`, and a curated list in `scala`). Later, we can include more and more features from `scala-cli` into `scala`. 

The sub-commands necessary to include in the `scala` to match the functionalities of current commands:
 - `compile`
 - `run`
 - `repl`
 - `doc`

> On top of that, we think that the following using-facing sub-commands should also be included:

Reviewer: ```suggestion
On top of that, we think that the following user-facing sub-commands should also be included:
```

 - `clean` - to rebuild project from start without any cached steps

>  - `setup-ide` - to control the setup for BSP for IDE integration

Reviewer: ```suggestion
 - `setup-ide` - to control the setup of BSP for IDE integration
```

 - `doctor` - to analyze if everything is installed properly

We also suggest to include additional sub-commands by default: 

>  - `fmt` - to format the code using scalafmt

Reviewer: ```suggestion
 - `fmt` - to format the code using `scalafmt`
```

Author: Not necessarily. It's scalafmt the project, not the command.

>  - `test` - to run test included in the code

Reviewer: ```suggestion
 - `test` - to run tests included in the code
```

 - `package` - to package the code into various package formats: JAR, "fat"  be, executable  be or even native application or docker image

>  - `shebang` - a command useful for scripting, designed to be included in `shebang` section of scripts

Reviewer: ```suggestion
 - `shebang` - a command useful for scripting, designed to be included in the `shebang` section of scripts
```

Author: We should also drop the backticks of the second "shebang".

>  - `shebang` - a command useful for scripting, designed to be included in the `shebang` section of scripts

Author: ```suggestion
 - `shebang` - a command useful for scripting, designed to be included in the "shebang" section of scripts
```

>  - `export` - transform current project to sbt / Mill project using the same settings as provided. Useful to evolve prototypes into bigger projects

Reviewer: ```suggestion
 - `export` - transform the current project to an `sbt` / `Mill` project using the same settings as provided. Useful to evolve prototypes into bigger projects
```

Author: No need of backticks around sbt or Mill here - it's the project names.

> Each of those commands expand what current scripts offer and can be discussed separately. We can even open a dedicated SIP for each of them.

Reviewer: ```suggestion
Each of these commands expand what the current scripts offer and can be discussed separately. We can even open a dedicated SIP for each of them.
```

Author: Should have been "expand*s*" too.

> Each of these commands expand what the current scripts offer and can be discussed separately. We can even open a dedicated SIP for each of them.

Author: ```suggestion
Each of these commands expands what the current scripts offer and can be discussed separately. We can even open a dedicated SIP for each of them.
```

Beyond that, `scala-cli` offers multiple sub-commands needed to manage itself (e.g. `update`) or its components (e.g. the Bloop server). In most cases,

> these are not user-facing, but these are handy. We can elaborate in more details what are those commands and why we need them if needed.

Reviewer: ```suggestion
these are not user-facing, but still handy. We can elaborate in more detail on what those commands are and why we need them, if necessary.
```

> Scala CLI can be also configured using ["using directives"](https://scala-cli.virtuslab.org/docs/guides/using-directives) - a comment-based configuration syntax that should be placed at the top of Scala files. This allows for self-containing examples within one file since most configuration can be provided either from the command line or via using directives (command line has precedence). This is a game changer for use cases like scripting, reproduction, or within the academic scope.

Reviewer: ```suggestion
Scala CLI can also be configured with ["using directives"](https://scala-cli.virtuslab.org/docs/guides/using-directives) - a comment-based configuration syntax that should be placed at the top of Scala files. This allows for self-containing examples within one file since most of the configuration can be provided either from the command line or via using directives (command line has precedence). This is a game changer for use cases like scripting, reproduction, or within the academic scope.
```

### Compatibility

> Adopting Scala CLI as the new `scala` command, as is, will change some the behaviour of today's scripts. Some examples:

Reviewer: ```suggestion
Adopting Scala CLI as the new `scala` command, as is, will change some of the behaviour of today's scripts. Some examples:
```

- `scala repl` needs to be run to start a REPL instead of just `scala`
- with `scala compile` and `scala doc`, some of the more obscure (or brand new) compile options need to be prefixed with `-O`
- Scala CLI recognizes tests based on the extension used (`*.test.scala`) so running `scala compile a.scala a.test.scala` will only compile `a.scala`
- Scala CLI has its own versioning scheme, that is not related to the Scala compiler. Default version used may dynamically change when new Scala version is released.
- By default, Scala CLI manages its own dependencies (e.g. scalac, zinc, Bloop) and resolves them lazily. This means that the first run of Scala CLI resolves quite some dependencies. Moreover, Scala CLI periodically checks for updates, new defaults accessing online resources (but it is not required to work, so Scala CLI can work in offline environment once setup)

> - Scala CLI can be also configured via using directives. Command line options have precendece over using directives, however using directives overrides defaults. Compiling a file starting with `//> using scala 2.13.8`, without providing a Scala version on the command line, will result in using `2.13.8` rather then default Scala version. We consider this a feature. However, technically, this is a breaking change.

Reviewer: ```suggestion
- Scala CLI can also be configured via using directives. Command line options have precedence over using directives, however using directives overrides defaults. Compiling a file starting with `//> using scala 2.13.8`, without providing a Scala version on the command line, will result in using `2.13.8` rather than the default Scala version. We consider this a feature. However, technically, this is a breaking change.
```

Author: "overrides" -> "override" too

> - Scala CLI can also be configured via using directives. Command line options have precedence over using directives, however using directives overrides defaults. Compiling a file starting with `//> using scala 2.13.8`, without providing a Scala version on the command line, will result in using `2.13.8` rather than the default Scala version. We consider this a feature. However, technically, this is a breaking change.

Author: ```suggestion
- Scala CLI can also be configured via using directives. Command line options have precedence over using directives, however using directives override defaults. Compiling a file starting with `//> using scala 2.13.8`, without providing a Scala version on the command line, will result in using `2.13.8` rather than the default Scala version. We consider this a feature. However, technically, this is a breaking change.
```

### Other concerns

Scala CLI brings [using directives](https://scala-cli.virtuslab.org/docs/guides/using-directives) and  [conventions to mark the test files](https://scala-cli.virtuslab.org/docs/commands/test#test-sources). We are not sure if both can be accepted as a part of this SIP or we should have seperate SIPs for both (we have opened a [pre-SIP for using directives](https://contributors.scala-lang.org/t/pre-sip-using-directives/5700/15))

> Scala CLI is an ambitious project, and may seem hard to maintain in long-run. // TODO

Reviewer: ```suggestion
Scala CLI is an ambitious project and may seem hard to maintain in the long-run. // TODO
```

## Alternatives

> Scala CLI has many alternatives. The most obvious ones are sbt, Mill, or other build tools. However, these are more complicated than Scala CLI, and what is more important they are not designed as command-line first tools. Ammonite, is another alternative, however it covers only part of the Scala CLI features (REPL and scripting), and lacks many of the Scala CLI features (incremental compilation, Scala version selection, support for Scala.js and Scala Native, just to name a few). 

Reviewer: ```suggestion
Scala CLI has many alternatives. The most obvious ones are `sbt`, `Mill` and other build tools. However, these are more complex than Scala CLI, and what is more important they are not designed as command-line first tools. Ammonite, is another alternative, however it covers only a subset of the features delivered by Scala CLI (REPL and scripting), and lacks many others (incremental compilation, Scala version selection, support for Scala.js and Scala Native, just to name a few). 
```

Author: No need of backticks around sbt and Mill here either.
